File: Examples/studentCollectionJsonRPC/StudentCollectionJsonRPC/src/java/server/Student.java
package sample.student.server;
/*
 * Copyright 2016 Tim Lindquist,
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Purpose: Example Java JsonRPC server for collection of students.
 *
 * Ser321 Distributed Apps, and Ser423 Mobile Apps
 * see http://pooh.poly.asu.edu/Mobile
 * @author Tim Lindquist Tim.Lindquist@asu.edu
 *         Software Engineering, CIDSE, IAFSE, ASU Poly
 * @version June 2016
 */

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.Vector;
import java.util.Arrays;

public class Student {

   private static final boolean debugOn = false;

   public String name;
   public int studentid;
   public Vector<String> takes;

   public Student(String name, int studentid, String[] courses){
      this.name = name;
      this.studentid = studentid;
      this.takes = new Vector<String>();
      this.takes.addAll(Arrays.asList(courses));
   }

   public Student(String jsonStr){
      try{
         JSONObject jo = new JSONObject(jsonStr);
         name = jo.getString("name");
         studentid = jo.getInt("studentid");
         takes = new Vector<String>();
         JSONArray ja = jo.optJSONArray("takes");
         for (int i=0; i< ja.length(); i++){
            takes.add(ja.getString(i));
         }
      }catch (Exception ex){
         System.out.println(this.getClass().getSimpleName()+
                            ": error converting from json string");
      }
   }

   public Student(JSONObject jsonObj){
      try{
         debug("constructor from json received: " + jsonObj.toString());
         name = jsonObj.optString("name","unknown");
         studentid = jsonObj.optInt("studentid",0);
         takes = new Vector<String>();
         JSONArray ja = jsonObj.getJSONArray("takes");
         for (int i=0; i< ja.length(); i++){
            takes.add(ja.getString(i));
         }
      }catch(Exception ex){
         System.out.println(this.getClass().getSimpleName()+
                            ": error converting from json string");
      }
   }

   public JSONObject toJson(){
      JSONObject jo = new JSONObject();
      try{
         jo.put("name",name);
         jo.put("studentid",studentid);
         jo.put("takes",takes);
      }catch (Exception ex){
         System.out.println(this.getClass().getSimpleName()+
                            ": error converting to json");
      }
      return jo;
   }

   public String toJsonString(){
      String ret = "";
      try{
         ret = this.toJson().toString();
      }catch (Exception ex){
         System.out.println(this.getClass().getSimpleName()+
                            ": error converting to json string");
      }
      return ret;
   }

   public String toString(){
      StringBuilder sb = new StringBuilder();
      sb.append("Student ").append(name).append(" has id ");
      sb.append(studentid).append(" and takes courses ");
      for (int i=0; i<takes.size(); i++){
         sb.append(takes.get(i)).append(" ");
      }
      return sb.toString();
   }

   private void debug(String message) {
      if (debugOn)
         System.out.println("debug: "+message);
   }

}

File: Examples/studentCollectionJsonRPC/StudentCollectionJsonRPC/src/cpp/server/Student.cpp
/**
 * Copyright (c) 2016 Tim Lindquist,
 * Software Engineering,
 * Arizona State University at the Polytechnic campus
 * <p/>
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation version 2
 * of the License.
 * <p/>
 * This program is distributed in the hope that it will be useful,
 * but without any warranty or fitness for a particular purpose.
 * <p/>
 * Please review the GNU General Public License at:
 * http://www.gnu.org/licenses/gpl-2.0.html
 * see also: https://www.gnu.org/licenses/gpl-faq.html
 * so you are aware of the terms and your rights with regard to this software.
 * Or, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,USA
 * <p/>
 * Purpose: C++ class to represent a student. This class is part of a
 * student collection distributed application that uses JsonRPC.
 * Meant to run on OSX, Debian Linux, and Raspberry Pi Debian.
 * <p/>
 * Ser321 Principles of Distributed Software Systems
 * see http://pooh.poly.asu.edu/Ser321
 * @author Tim Lindquist (Tim.Lindquist@asu.edu),ASU-IAFSE,Software Engineering
 * @file    Student.cpp
 * @date    July, 2016
 * @license See above
 **/

#include "Student.hpp"
#include <iostream>
#include <stdlib.h>
Student::Student(){
   name = "";
   studentid = 0;
   takes = vector<string>();
}

Student::Student(string aName, int aStudentId, vector<string> aTakes) {
   name = aName;
   studentid = aStudentId;
   takes = aTakes;
}

Student::Student(const Json::Value& jsonObj){
   string nameStr = "name";
   string studentidStr = "studentid";
   string takesStr = "takes";
   Json::Value::Members mbr = jsonObj.getMemberNames();
   for(vector<string>::const_iterator i = mbr.begin(); i!= mbr.end(); i++){
      Json::Value jsonM = jsonObj[*i];
      if(nameStr.compare(*i)==0){
         name = jsonM.asString();
      }else if(studentidStr.compare(*i)==0){
         studentid = jsonM.asInt();
      }else if(takesStr.compare(*i)==0){
         takes = vector<string>();
         for(int i=0; i<jsonM.size(); i++){
            takes.push_back(jsonM[i].asString());
         }
      }
   }
}

Student::Student(string jsonString){
   string nameStr = "name";
   string studentidStr = "studentid";
   string takesStr = "takes";
   Json::Reader reader;
   Json::Value root;
   bool parseSuccess = reader.parse(jsonString,root,false);
   if(parseSuccess){
      Json::Value::Members mbr = root.getMemberNames();
      for(vector<string>::const_iterator i = mbr.begin(); i!= mbr.end(); i++){
         Json::Value jsonM = root[*i];
         if(nameStr.compare(*i)==0){
            name = jsonM.asString();
         }else if(studentidStr.compare(*i)==0){
            studentid = jsonM.asInt();
         }else if(takesStr.compare(*i)==0){
            takes = vector<string>();
            for(int i=0; i<jsonM.size(); i++){
               takes.push_back(jsonM[i].asString());
            }
         }
      }
   }else{
      cout << "Student constructor parse error with input: " << jsonString
           << endl;
   }
}

Student::~Student() {
   name = "";
   studentid = -1;
}

string Student::toJsonString(){
   string ret = "{}";
   Json::Value jsonLib;
   jsonLib["name"] = name;
   jsonLib["studentid"] = studentid;
   Json::Value tmp(Json::arrayValue);
   for(int i=0; i<takes.size(); i++){
      tmp[i]=takes[i];
   }
   jsonLib["takes"] = tmp;
   ret = jsonLib.toStyledString();
   return ret;
}

Json::Value Student::toJson(){
   Json::Value jsonLib;
   jsonLib["name"] = name;
   jsonLib["studentid"] = studentid;
   Json::Value tmp(Json::arrayValue);
   for(int i=0; i<takes.size(); i++){
      tmp[i]=takes[i];
   }
   jsonLib["takes"] = tmp;
   return jsonLib;
}

void Student::setValues(string aName, int aStudentId, vector<string> aTakes) {
   name = aName;
   studentid = aStudentId;
   takes = aTakes;
}

void Student::print(){
   cout << "Student " << name << " with ID "
        << studentid << " takes ";
   for (int i=0; i<takes.size(); i++){
      cout << takes[i] << " ";
   }
   cout << endl;
}
